UI/MapWnd.cpp

```diff
@@ -2778,57 +2781,147 @@ void MapWnd::ClearSystemRenderingBuffers() {
+        const SupplyLaneMMap& supply_lanes)
+    {
+        good_path.clear();
+
+        // No terminal points, so all paths lead nowhere.
+        if (terminal_points.empty())
+            return;
+
+        boost::unordered_map<int, AncestorInfo> ancestors;
+        std::deque<PrevCurrInfo> tryNext;
+        std::vector<int> path;
+
+        // A list of all reachable midpoints between at least two different terminal points
+        std::vector<int> reachable_midpoints;
+
+        // Start from all terminal points.
+        for (boost::unordered_set<int>::const_iterator start_it = terminal_points.begin();
+             start_it != terminal_points.end(); ++start_it) {
+            tryNext.push_back(PrevCurrInfo(*start_it, *start_it, *start_it));
+            ancestors.insert(std::make_pair(*start_it, AncestorInfo(*start_it, *start_it)));
+        }
+
+        // Find all reachable midpoints where paths from two different
+        // terminal points meet.
+        while (!tryNext.empty() ) {
+            const PrevCurrInfo & curr = tryNext.front();
+
+            std::pair<SupplyLaneMMap::const_iterator, SupplyLaneMMap::const_iterator> supplylane_endpoints
+                = supply_lanes.equal_range(curr.curr);
+            for (SupplyLaneMMap::const_iterator sup_it = supplylane_endpoints.first;
+                 sup_it != supplylane_endpoints.second; ++sup_it)
+            {
+                int next = sup_it->second;
+                if (next == curr.prev)
+                    continue;
+                boost::unordered_map<int, AncestorInfo>::iterator ancestor = ancestors.find(next);
+
+                // Unvisited system, so create new ancestor.
+                if (ancestor == ancestors.end()) {
+                    ancestors.insert(std::make_pair(next, AncestorInfo(curr.curr, curr.origin)));
+                    tryNext.push_back(PrevCurrInfo(curr.curr, next, curr.origin));
+                }
+
+                // Previously visited so modify older ancestor to merge
+                // paths/create end points.
+                else {
+
+                    // Same origin so add to ancestors.
+                    if (ancestor->second.single_origin
+                        && ancestor->second.single_origin == curr.origin) {
+                        ancestor->second.ids.push_back(curr.curr);
                     }
-                    return path;
+
+                    // Different origins so mark both as ends of good paths
+                    else if (ancestor->second.single_origin
+                        && ancestor->second.single_origin != curr.origin) {
+                        // Single origin becomes multi origin and these
+                        // points are marked as midpoints.
+                        ancestor->second.single_origin = boost::none;
+                        reachable_midpoints.push_back(curr.curr);
+                        reachable_midpoints.push_back(next);
+                    }
```

Comment:
same here... make it clear there's an else block by not having space between the end of the if {} and the else